torch/nn/modules/module.py

```diff
@@ -1278,8 +1278,47 @@ def _save_to_state_dict(self, destination, prefix, keep_vars):
         if getattr(self.__class__, "get_extra_state", Module.get_extra_state) is not Module.get_extra_state:
             destination[extra_state_key] = self.get_extra_state()
 
-    # The user can pass an optional arbitrary mappable object to `state_dict`, in which case `state_dict` returns
-    # back that same object. But if they pass nothing, an `OrederedDict` is created and returned.
+    def _state_dict_impl(self, destination, prefix, keep_vars):
+        r"""Holds the actual implementation of
+        :meth:`~torch.nn.Module.state_dict`, with recursive calls for
+        descendants of this module.
+
+        In rare cases, users can call this directly to provide a custom
+        :attr:`destination`.
+
+        Args:
+            destination (dict): a dict where state will be stored
+            prefix (str): the prefix for parameters and buffers used in this
+                module
+            keep_vars (bool): whether NOT to return buffers detached from
+                autograd
+        """
+        destination._metadata[prefix[:-1]] = local_metadata = dict(version=self._version)
+        self._save_to_state_dict(destination, prefix, keep_vars)
+        for name, module in self._modules.items():
+            if module is not None:
+                module.state_dict(destination, prefix + name + '.', keep_vars=keep_vars)
+        for hook in self._state_dict_hooks.values():
+            hook_result = hook(self, destination, prefix, local_metadata)
+            if hook_result is not None:
+                destination = hook_result
+        return destination
+
+    # TODO: state dict should be Mapping[str, Any] (OrderedDict[str, Any])
+    # https://github.com/pytorch/pytorch/issues/72707#issuecomment-1040487230
+
+    # TODO: remove overloads when `destination` is removed from `state_dict`
+    # Keep only the second signature after removal
+    # The first overload is only for BC
+
+    # Comment on the comment below: it originated from the type stub in .pyi.
```

Comment:
The comments below are not really needed.

Reply:
will fix